utils/wait.ahk

```diff
@@ -59,3 +59,55 @@ WaitForVictoryOrDefeat() {
         Sleep(2000)
     }
 }
+
+WaitForUpgrade1() {
+    while true {
+        if SearchUpgrade1() {
+            return
+        }
+        if defeated or SearchImage("states\defeat") {
+            global defeated := true
+            return
+        }
+        CheckLevelUp()
+    }
+}
+
+WaitForUpgrade2() {
+    while true {
+        if SearchUpgrade2() {
+            return
+        }
+        if defeated or SearchImage("states\defeat") {
+            global defeated := true
+            return
+        }
+        CheckLevelUp()
+    }
+}
+
+WaitForUpgrade3() {
+    while true {
+        if SearchUpgrade3() {
+            return
+        }
+        if defeated or SearchImage("states\defeat") {
+            global defeated := true
+            return
+        }
+        CheckLevelUp()
+    }
+}
+
+Wait(delay) {
+    if defeated {
+        return
+    }
+    Sleep(delay)
+    if SearchImage("states\defeat") {
+        Sleep(500)
+        ClickImage("buttons\home_defeat", 2000)
+        global defeats := defeats + 1
+        LogMsg("Defeat")
+    }
```

Comment:
Hardcoded delays should not be used in game scripts at all, the game doesn't always run at the same speed and this will cause problems.

Reply:
In that case, we might need some other way to recognize
- when you can buy certain towers (eg. your hero)
- when a specific type of bloon starts spawning
- when bloons reach a certain part of the map
As this is what I used this function for.

Reply:
That's definitely going to be a challenge but older versions used hardcoded delays with an adjustable multiplier in the configs and even that caused a lot of issues. The largest difference I have seen has been the game running about 25 % slower.

Reply:
I will try working on the "bloons reach specific parts of the map" next. 
Being able to detect when you can buy the hero doesn't seem like it would be that hard, as their picture is always guaranteed to be on screen. Super Monkeys on the other hand have nothing close to their price category, so they will always be hard to detect.